[PE-1926] validation bug for radio and/or checkbox elements

mattpepper opened the pull request:
# [PE-1926] validation bug for radio and/or checkbox elements

## Issue
Bug exposed by grouped form elements when part of group is hidden

## Current behaviour:
When a form validation error is occurs where:
1. the form is bound to generic validation via the `.js-form` selector
2. and some form elements are `:hidden`;
3. the `:hidden` elements value attribute is "cleared" via $(*element*).val('') 

This behaviour is incorrect for both `radio` and/or `checkbox` elements where the value should **not** be cleared.

### Example:
In the shown example, the form:
  * validates that the required number of 'checked' elements in the group is correct
  * has two `radio` elements in the group, which are `:hidden` by a `checkbox` toggle element

 When the form is submitted:
 1. without the required number of elements selected
 2. *and* the `radio` elements are `:hidden`
 3. the bug results in:
    a. the `radio:hidden` element values being set to '' (empty)
    b. *and*, as `flushHiddenElementErrors` uses the `[name="..."]` selector (binding the entire group),
         the visible `checkbox` element values are *also* being set to '' (empty)

## Solution
This change will prevent the value being set to '' (empty) on `radio:hidden` or `checkbox:hidden` elements.

![PE-1926 validation bug for radio and/or checkbox elements](https://cloud.githubusercontent.com/assets/5468091/14892031/06ac1316-0d61-11e6-864a-e28ae090d4aa.gif)

mattpepper commented:
FYI - PR closed to add some edge cases to the form's acceptance tests.

PR re-opened without changes.

feedmypixel commented on assets/javascripts/validation/form.js:
```diff
@@ -122,8 +122,12 @@ var flushHiddenElementErrors = function (invalidInputs) {
+        $elem.closest('.form-field-group').removeClass('form-field-group--error');
+      }
     }
   }
```
@mattpepper for the conditional something like the following feels a little clearer to me, then you don't need the comment:

```
var inputType = $elem.prop('type');

if (inputType !== 'radio' && inputType !== 'checkbox') {
   $elem.val('');
   $elem.closest('.form-field-group').removeClass('form-field-group--error');
}
```

Can you also add a test around this for assets-frontend please in https://github.com/hmrc/assets-frontend/blob/master/assets/test/specs/form.spec.js

feedmypixel commented:
A comment added - from what I can see this makes sense as the original implementation was to reset a `text` inputs value. 
This is specifically being used on 2 Step Verification so it may be worth asking @benaveryee to check out the branch and test on the enter mobile page for that journey.

mattpepper commented:
@feedmypixel : Addressed yu comments and added tests
@benaveryee: please could you check this out on 2 step verification and test as @feedmypixel  suggested?

benaveryee commented:
2 Step Verification tests are fine running against this branch of assets-frontend :+1:

feedmypixel commented on assets/javascripts/validation/form.js:
```diff
@@ -117,15 +117,21 @@ var createErrorSummaryListItem = function ($liElement, validator, errorDetail) {
-      $elem.closest('.form-field-group').removeClass('form-field-group--error');
+
+      // filter out radios and checkboxes before clearing the element value   
+      if (inputType !== 'radio' && inputType !== 'checkbox') {
```
so @mattpepper after discussion this `if` is not needed please remove it.

mattpepper replied:
to confirm - remove the `if` type, no the :hidden `if`?

feedmypixel replied:
> to confirm - remove the if type, not the :hidden if?

`if (inputType !== 'radio' && inputType !== 'checkbox') {`

feedmypixel commented on assets/javascripts/validation/form.js:
```diff
@@ -117,15 +117,21 @@ var createErrorSummaryListItem = function ($liElement, validator, errorDetail) {
+
+      // filter out radios and checkboxes before clearing the element value   
+      if (inputType !== 'radio' && inputType !== 'checkbox') {
+        $elem.val('');
```
@mattpepper this `$elem.val('');` is also not needed.

rpowis replied:
Essentially the validation module shouldn't be concerned with clearing form data, only validating it

mattpepper replied:
Wholly agree with this - values which don't validate should not be cleared. 👍

feedmypixel commented:
@mattpepper with tests all you need to do is check that when an item is hidden that the `form-field-group--error` class is removed. It is a little bit of crossover but I feel a simple test to check this is all you need here. No need for the tests you have done and all the fixture work.

feedmypixel commented:
@benaveryee the only thing with removing the clearing of inputs here in form validation means that you will have to replace the toggle work found here:
[OTPFE - enterMobileNumber](https://github.tools.tax.service.gov.uk/HMRC/one-time-password-frontend/blob/29712d35de639da4fa5d4e69fa7cb2a1ff2c09ac/app/views/otpfe/registration/enterMobileNumber.scala.html#L70)
and here:
[OTPFE - enterLandlineNumber](https://github.tools.tax.service.gov.uk/HMRC/one-time-password-frontend/blob/29712d35de639da4fa5d4e69fa7cb2a1ff2c09ac/app/views/otpfe/registration/enterLandlineNumber.scala.html#L70)

With [toggleDynamicFormFields](https://github.com/hmrc/assets-frontend/blob/master/assets/javascripts/modules/toggleDynamicFormFields.js) as the form validation work will no longer clear input values when something is hidden.

mattpepper commented:
@feedmypixel form flushHiddenElementErrors() updated as per comments. Tests & fixture stripped down as suggested.

rpowis commented on assets/test/specs/fixtures/form-fixture.html:
```diff
@@ -1,51 +1,78 @@
-             id="radio-yes"
-             value="yes"
-             name="radio-example"
+<div class="js"><!-- required to hookup css for dynamic field toggle -->
```
Could this not be added by calling `document.body.classList.add("js")` in test setup?

rpowis replied:
Removing this wrapping div will also make the diff clearer

mattpepper replied:
updated 👌

rpowis commented on assets/test/specs/fixtures/form-fixture.html:
```diff
@@ -1,51 +1,78 @@
-             value="yes"
-             name="radio-example"
+<div class="js"><!-- required to hookup css for dynamic field toggle -->
+  <form class="js-form" autocomplete="off" novalidate="novalidate"  data-dynamic-form="true">
```
`data-dynamic-form` doesn't need a value of true. Just the existence of the attribute makes it true.

mattpepper replied:
updated 👌

feedmypixel commented on assets/test/specs/fixtures/form-fixture.html:
```diff
@@ -46,6 +46,31 @@ <h2 id="errorSummaryHeading" class="h3-heading">There are errors on the page.</h
-
+  
+  <fieldset class="form-field-group">
+    <span id="radio-group-error-message" class="error-notification" role="tooltip" data-input-name="radio-group-example">
```
`data-input-name="radio-group-example"` isn't needed right?

rpowis commented on assets/test/specs/form.spec.js:
```diff
@@ -32,6 +40,8 @@ describe('Form Validation', function () {
     loadFixtures('form-fixture.html');
     form = require('../../javascripts/validation/form.js');
+    toggleDynamic = require('../../javascripts/modules/toggleDynamicFormFields.js');
+    document.body.classList.add("js")
```
This class actually needs to be added to the fixture container, not the body, sorry:

`document.getElementById(jasmine.getFixtures().containerId).classList.add('js');`

rpowis replied:
When it's on the body it borks other tests because it isn't being removed between runs of suites. Fixtures are destroyed with each run so it isolates you tests.

mattpepper replied:
done 👍

feedmypixel commented on assets/test/specs/fixtures/form-fixture.html:
```diff
@@ -46,6 +46,31 @@ <h2 id="errorSummaryHeading" class="h3-heading">There are errors on the page.</h
+    </div>
+  </fieldset>
+  
   <button id="submit-button" type="submit">Submit</button>
```
To make it clearer css selectors, id's and attributes I would only include what is needed here. You don't need semantic html for fixtures.

mattpepper replied:
Just a "cut and paste" from view - I understand the point and will update.

mattpepper replied:
Cleaned up the selectors - the addition to fixture is as selector-verbose as the existing html 👌

feedmypixel commented on assets/test/specs/form.spec.js:
```diff
@@ -24,6 +29,9 @@ var setup = function () {
   $radioYesElement = $('#radio-yes');
   $radioNoElement = $('#radio-no');
+  $radioGroupThis = $('#radio-this');
+  $radioGroupThat = $('#radio-that');
```
radio what? not sure what `this` and `that` means here

mattpepper replied:
These are token values - is the question should they be `$radioGroupThisElement` or that they be like those already used, i.e. `$radioGroupNoElement`?

rpowis replied:
I think the issue is that `this` and `that` are ambiguous and you need to refer to the fixture to see what they are. Can you rename them to be more descriptive?

mattpepper replied:
done

feedmypixel commented on assets/test/specs/form.spec.js:
```diff
@@ -220,5 +229,41 @@ describe('Form Validation', function () {
+          expect($errorMessages.eq(e).text()).not.toBe($radioNoGroupElement.attr('data-msg-required'));
+        }
+      });
+    });
```
I would also add a further describe
describe('When I open the grouped input elements, hit submit and cause an error, then close the grouped input elements');
it('The error should not be visible');
it('The form should have the grouped radio required error message');

**Note you will need to add something along the lines of https://github.com/hmrc/assets-frontend/blob/PE-1926-validation-bug/assets/test/specs/form.spec.js#L94-L97 to stop the page refreshing on successful submit**

mattpepper replied:
done 👍

feedmypixel commented on assets/test/specs/form.spec.js:
```diff
@@ -220,5 +229,41 @@ describe('Form Validation', function () {
+
+      it('The form should have the grouped radio required error message', function () {
+        var $errorMessages = $errorSummaryMessages.find('> li');
+        expect($errorMessages.text().indexOf($radioYesGroupElement.attr('data-msg-required'))).not.toBe(-1);
```
This feels a little simpler? and is being used elsewhere in the file?
`expect($errorMessages.text()).toBe($radioYesGroupElement.attr('data-msg-required'));`

mattpepper replied:
done 👍

feedmypixel commented:
LGTM 👍  - thanks @mattpepper
